File: multilibrary/src/main/java/com/zhongjh/albumcamerarecorder/album/ui/mediaselection/MediaViewUtil.java
package com.zhongjh.albumcamerarecorder.album.ui.mediaselection;

import android.util.Log;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.zhongjh.albumcamerarecorder.R;
import com.zhongjh.albumcamerarecorder.listener.OnLogListener;
import com.zhongjh.albumcamerarecorder.model.SelectedModel;
import com.zhongjh.albumcamerarecorder.album.entity.Album2;
import com.zhongjh.albumcamerarecorder.album.widget.recyclerview.RecyclerLoadMoreView;
import com.zhongjh.albumcamerarecorder.settings.GlobalSpec;
import com.zhongjh.common.entity.LocalMedia;
import com.zhongjh.albumcamerarecorder.model.MainModel;
import com.zhongjh.albumcamerarecorder.album.ui.mediaselection.adapter.AlbumAdapter;
import com.zhongjh.albumcamerarecorder.album.utils.UiUtils;
import com.zhongjh.albumcamerarecorder.album.ui.mediaselection.adapter.widget.MediaGridInset;
import com.zhongjh.albumcamerarecorder.settings.AlbumSpec;

/**
 * 以前是MediaSelectionFragment,现在为了能滑动影响上下布局，放弃Fragment布局，直接使用RecyclerView
 * Fragment嵌套RecyclerView的话，会在配置小的机器下产生性能卡顿问题
 *
 * @author zhongjh
 * @date 2022/9/19
 */
public class MediaViewUtil implements
        AlbumAdapter.CheckStateListener, AlbumAdapter.OnMediaClickListener {

    public MediaViewUtil(FragmentActivity activity,
                         Fragment fragment,
                         MainModel mainModel,
                         SelectedModel selectedModel,
                         RecyclerLoadMoreView recyclerView,
                         AlbumAdapter.CheckStateListener checkStateListener,
                         AlbumAdapter.OnMediaClickListener onMediaClickListener) {
        mActivity = activity;
        mFragment = fragment;
        mMainModel = mainModel;
        mSelectedModel = selectedModel;
        mRecyclerView = recyclerView;
        mCheckStateListener = checkStateListener;
        mOnMediaClickListener = onMediaClickListener;
        init();
    }

    private final FragmentActivity mActivity;
    private final Fragment mFragment;
    private final MainModel mMainModel;
    private final SelectedModel mSelectedModel;
    private final RecyclerLoadMoreView mRecyclerView;
    private AlbumAdapter mAdapter;
    private Album2 mAlbum;
    private AlbumSpec mAlbumSpec;
    /**
     * 单选事件
     */
    private AlbumAdapter.CheckStateListener mCheckStateListener;
    /**
     * 点击事件
     */
    private AlbumAdapter.OnMediaClickListener mOnMediaClickListener;

    private void init() {
        // 先设置recyclerView的布局
        int spanCount;
        mAlbumSpec = AlbumSpec.INSTANCE;
        if (mAlbumSpec.getGridExpectedSize() > 0) {
            spanCount = UiUtils.spanCount(mActivity, mAlbumSpec.getGridExpectedSize());
        } else {
            spanCount = mAlbumSpec.getSpanCount();
        }
        // 删除动画
        mRecyclerView.setItemAnimator(null);
        mRecyclerView.setLayoutManager(new GridLayoutManager(mActivity, spanCount));
        // 需要先设置布局获取确定的spanCount，才能设置adapter
        mAdapter = new AlbumAdapter(mActivity, mSelectedModel, getImageResize());
        Log.d("onSaveInstanceState", " mAdapter");
        mAdapter.registerCheckStateListener(this);
        mAdapter.registerOnMediaClickListener(this);
        mAdapter.setHasStableIds(true);
        mRecyclerView.setHasFixedSize(true);

        // 加载线，recyclerView加载数据
        int spacing = mActivity.getResources().getDimensionPixelSize(R.dimen.z_media_grid_spacing);
        mRecyclerView.addItemDecoration(new MediaGridInset(spanCount, spacing, false));
        mRecyclerView.setAdapter(mAdapter);

        // 加载更多事件
        mRecyclerView.setOnRecyclerViewLoadMoreListener(() -> mMainModel.addAllPageMediaData(mAlbum.getId(), mAlbumSpec.getPageSize()));

        // 监听到新的相册数据
        mMainModel.getLocalMediaPages().observe(mFragment.getViewLifecycleOwner(), mediaData -> {
            // 如果没有数据，则关闭下拉加载
            mRecyclerView.setEnabledLoadMore(!mediaData.getData().isEmpty());
            if (mMainModel.getPage() == 1) {
                mAdapter.setData(mediaData.getData());
                mRecyclerView.scrollToPosition(0);
            } else {
                mAdapter.addData(mediaData.getData());
            }
        });

        // 输出失败信息
        mMainModel.getOnFail().observe(mFragment.getViewLifecycleOwner(), throwable -> {
            if (GlobalSpec.INSTANCE.getOnLogListener() != null) {
                GlobalSpec.INSTANCE.getOnLogListener().logError(throwable);
            }
        });
    }

    public void onDestroyView() {
        mAdapter.unregisterCheckStateListener();
        mAdapter.unregisterOnMediaClickListener();
        mCheckStateListener = null;
        mOnMediaClickListener = null;
        mAdapter = null;
    }

    /**
     * 重新获取数据源
     */
    public void reloadPageMediaData() {
        mMainModel.reloadPageMediaData(mAlbum.getId(), mAlbumSpec.getPageSize());
    }

    /**
     * 每次筛选后，重新查询数据
     *
     * @param album 专辑
     */
    public void load(Album2 album) {
        mAlbum = album;
        reloadPageMediaData();
    }

    /**
     * 刷新数据源
     */
    public void refreshMediaGrid() {
        mAdapter.notifyItemRangeChanged(0, mAdapter.getItemCount());
    }

    /**
     * 刷新列表数据
     */
    public void notifyItemByLocalMedia() {
        mAdapter.notifyCheckStateChanged();
    }

    @Override
    public void onUpdate() {
        // 通知外部活动检查状态改变
        if (mCheckStateListener != null) {
            mCheckStateListener.onUpdate();
        }
    }

    @Override
    public void onMediaClick(Album2 album, ImageView imageView, LocalMedia item, int adapterPosition) {
        if (mOnMediaClickListener != null) {
            mOnMediaClickListener.onMediaClick(mAlbum, imageView,
                    item, adapterPosition);
        }
    }

    /**
     * 返回图片调整大小
     *
     * @return 列表的每个格子的宽度 * 缩放比例
     */
    private int getImageResize() {
        int imageResize;
        RecyclerView.LayoutManager lm = mRecyclerView.getLayoutManager();
        int spanCount = 0;
        if (lm != null) {
            spanCount = ((GridLayoutManager) lm).getSpanCount();
        }
        int screenWidth = mActivity.getResources().getDisplayMetrics().widthPixels;
        int availableWidth = screenWidth - mActivity.getResources().getDimensionPixelSize(
                R.dimen.z_media_grid_spacing) * (spanCount - 1);
        // 图片调整后的大小：获取列表的每个格子的宽度
        imageResize = availableWidth / spanCount;
        // 图片调整后的大小 * 缩放比例
        imageResize = (int) (imageResize * AlbumSpec.INSTANCE.getThumbnailScale());
        return imageResize;
    }
}

File: multilibrary/src/main/java/com/zhongjh/multimedia/album/ui/mediaselection/MediaViewUtil.kt
package com.zhongjh.multimedia.album.ui.mediaselection

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.GestureDetector
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.widget.ImageView
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.zhongjh.common.entity.LocalMedia
import com.zhongjh.multimedia.R
import com.zhongjh.multimedia.album.entity.Album
import com.zhongjh.multimedia.album.ui.mediaselection.adapter.AlbumAdapter
import com.zhongjh.multimedia.album.ui.mediaselection.adapter.widget.MediaGridInset
import com.zhongjh.multimedia.album.utils.UiUtils.spanCount
import com.zhongjh.multimedia.album.widget.recyclerview.RecyclerLoadMoreView
import com.zhongjh.multimedia.album.widget.recyclerview.listener.OnRecyclerViewScrollStateListener
import com.zhongjh.multimedia.model.MainModel
import com.zhongjh.multimedia.model.SelectedModel
import com.zhongjh.multimedia.settings.AlbumSpec
import com.zhongjh.multimedia.settings.AlbumSpec.thumbnailScale
import com.zhongjh.multimedia.settings.GlobalSpec.imageEngine
import com.zhongjh.multimedia.settings.GlobalSpec.onLogListener
import com.zhongjh.multimedia.utils.LifecycleFlowCollector

/**
 * 以前是MediaSelectionFragment,现在为了能滑动影响上下布局，放弃Fragment布局，直接使用RecyclerView
 * Fragment嵌套RecyclerView的话，会在配置小的机器下产生性能卡顿问题
 *
 * @param checkStateListener 单选事件
 * @param onMediaClickListener 点击事件
 *
 * @author zhongjh
 * @date 2022/9/19
 */
class MediaViewUtil(
    private val context: Context, private val owner: LifecycleOwner, private val mainModel: MainModel, private val selectedModel: SelectedModel,
    private val recyclerView: RecyclerLoadMoreView, private val placeholder: Drawable?,
    private var checkStateListener: AlbumAdapter.CheckStateListener?, private var onMediaClickListener: AlbumAdapter.OnMediaClickListener?
) : AlbumAdapter.CheckStateListener, AlbumAdapter.OnMediaClickListener {

    private var mAdapter: AlbumAdapter? = null
    private lateinit var mAlbum: Album
    private var mAlbumSpec = AlbumSpec

    init {
        init()
    }

    private fun init() {
        // 先设置recyclerView的布局
        val spanCount = if (mAlbumSpec.gridExpectedSize > 0) {
            spanCount(context, mAlbumSpec.gridExpectedSize)
        } else {
            mAlbumSpec.spanCount
        }
        // 删除动画
        recyclerView.itemAnimator = null
        recyclerView.layoutManager = GridLayoutManager(context, spanCount)
        // 需要先设置布局获取确定的spanCount，才能设置adapter
        mAdapter = AlbumAdapter(selectedModel, placeholder, imageResize)
        mAdapter?.registerCheckStateListener(this)
        mAdapter?.registerOnMediaClickListener(this)
        mAdapter?.setHasStableIds(true)
        recyclerView.setHasFixedSize(true)

        // 加载线,如果有配置则加载
        if (mAlbumSpec.spacing > 0) {
            recyclerView.addItemDecoration(MediaGridInset(spanCount, mAlbumSpec.spacing, false))
        } else {
            val spacing = context.resources.getDimensionPixelSize(R.dimen.z_media_grid_spacing)
            recyclerView.addItemDecoration(MediaGridInset(spanCount, spacing, false))
        }

        // recyclerView加载数据
        recyclerView.adapter = mAdapter

        // 添加长按滑动选择功能
        setupLongPressSlideSelection()

        // 加载更多事件
        recyclerView.setOnRecyclerViewLoadMoreListener { mainModel.addAllPageMediaData(mAlbum.id, mAlbumSpec.pageSize) }

        // 滑动事件
        recyclerView.setOnRecyclerViewScrollStateListener(object : OnRecyclerViewScrollStateListener {
            override fun onScrollFast() {
                imageEngine.pauseRequests(context)
            }

            override fun onScrollSlow() {
                imageEngine.resumeRequests(context)
            }
        })

        LifecycleFlowCollector.collect(owner, mainModel.mediaPageState) { state ->
            when (state) {
                // 监听到新的相册数据
                is MainModel.MediaPageState.RefreshSuccess -> {
                    // 如果没有数据，则关闭下拉加载
                    recyclerView.setEnabledLoadMore(state.refreshMediaData.data.isNotEmpty())
                    mAdapter?.setReloadPageMediaData(state.refreshMediaData)
                    recyclerView.scrollToPosition(0)
                }

                // 监听到下拉相册数据
                is MainModel.MediaPageState.LoadMoreSuccess -> {
                    // 如果没有数据，则关闭下拉加载
                    recyclerView.setEnabledLoadMore(state.startPosition != -1)
                    mAdapter?.setData(state.data)
                    mAdapter?.notifyItemInserted(state.startPosition)
                }

                // 输出失败信息
                is MainModel.MediaPageState.Error -> {
                    onLogListener?.logError(state.cause)
                }

                else -> {}
            }
        }
    }

    fun onDestroyView() {
        mAdapter?.unregisterCheckStateListener()
        mAdapter?.unregisterOnMediaClickListener()
        checkStateListener = null
        onMediaClickListener = null
        mAdapter = null
    }

    /**
     * 每次筛选后，重新查询数据
     *
     * @param album 专辑
     */
    fun load(album: Album) {
        mAlbum = album
        mainModel.reloadPageMediaData(mAlbum.id, mAlbumSpec.pageSize)
    }

    /**
     * 刷新数据源
     */
    fun refreshMediaGrid() {
        mAdapter?.let { adapter ->
            adapter.notifyItemRangeChanged(0, adapter.itemCount)
        }
    }

    /**
     * 刷新列表数据
     */
    fun notifyItemByLocalMedia(position: Int) {
        mAdapter?.notifyCheckStateChanged(position)
    }

    /**
     * 设置长按滑动选择功能
     */
    private fun setupLongPressSlideSelection() {
        var isSelecting = false
        var lastSelectedPosition = -1

        // 创建手势检测器处理长按事件
        val gestureDetector = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onLongPress(e: MotionEvent) {
                val child = recyclerView.findChildViewUnder(e.x, e.y)
                if (child != null) {
                    val position = recyclerView.getChildAdapterPosition(child)
                    if (position != RecyclerView.NO_POSITION) {
                        isSelecting = true
                        lastSelectedPosition = position
                        toggleSelection(child.findViewById(R.id.media_thumbnail), position)

                        // 添加长按激活选择模式的震动反馈
                        child.performHapticFeedback(
                            // 长按反馈（较强震动）
                            HapticFeedbackConstants.LONG_PRESS,
                            HapticFeedbackConstants.FLAG_IGNORE_VIEW_SETTING
                        )
                    }
                }
                super.onLongPress(e)
            }

            // 添加此方法确保触摸事件能传递到OnItemTouchListener
            override fun onDown(e: MotionEvent): Boolean {
                // 允许所有触摸事件传递
                return true
            }
        })

        // 将setOnTouchListener替换为addOnItemTouchListener
        recyclerView.addOnItemTouchListener(object : RecyclerView.OnItemTouchListener {
            override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
                // 先让GestureDetector处理事件（长按检测）
                gestureDetector.onTouchEvent(e)

                // 仅在长按激活选择模式后拦截事件 拦截所有事件，交由onTouchEvent处理
                return isSelecting
            }

            override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {
                when (e.action) {
                    MotionEvent.ACTION_MOVE -> {
                        if (isSelecting) {
                            val child = recyclerView.findChildViewUnder(e.x, e.y)
                            if (child != null) {
                                val position = recyclerView.getChildAdapterPosition(child)
                                if (position != RecyclerView.NO_POSITION && position != lastSelectedPosition) {
                                    lastSelectedPosition = position
                                    toggleSelection(child.findViewById(R.id.media_thumbnail), position)
                                }
                            }
                        }
                    }
                    MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                        isSelecting = false
                        lastSelectedPosition = -1
                    }
                }
            }
            override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}
        })
    }

    /**
     * 切换项目选择状态
     */
    private fun toggleSelection(imageView: ImageView, position: Int) {
        val media = mAdapter?.getItem(position) ?: return
        mAdapter?.onCheckViewClicked(imageView, media, context, position)
    }

    override fun onUpdate() {
        // 通知外部活动检查状态改变
        checkStateListener?.onUpdate()
    }

    override fun onMediaClick(album: Album?, imageView: ImageView?, item: LocalMedia?, adapterPosition: Int) {
        onMediaClickListener?.onMediaClick(mAlbum, imageView, item, adapterPosition)
    }

    private val imageResize: Int
        /**
         * 返回图片调整大小
         *
         * @return 列表的每个格子的宽度 * 缩放比例
         */
        get() {
            var imageResize: Int
            val lm = recyclerView.layoutManager
            var spanCount = 0
            if (lm != null) {
                spanCount = (lm as GridLayoutManager).spanCount
            }
            val screenWidth = context.resources.displayMetrics.widthPixels
            val availableWidth = screenWidth - context.resources.getDimensionPixelSize(
                R.dimen.z_media_grid_spacing
            ) * (spanCount - 1)
            // 图片调整后的大小：获取列表的每个格子的宽度
            imageResize = availableWidth / spanCount
            // 图片调整后的大小 * 缩放比例
            imageResize = (imageResize * thumbnailScale).toInt()
            return imageResize
        }
}
